File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ShouldImplementExportedInterfaces.cs
using System;
using System.ComponentModel.Composition;

namespace Classes
{
    static class Constants
    {
        public const string ContractName = "asdasd";
    }

    interface MyInterface { }

    [Export(typeof(MyInterface))] // Noncompliant {{Implement 'MyInterface' on 'NotExported' or remove this export attribute.}}
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    [Export(typeof(Exported))] // Noncompliant {{Derive from 'Exported' on 'NotExported' or remove this export attribute.}}
//   ^^^^^^^^^^^^^^^^^^^^^^^^
    class NotExported
    {
    }

    [Export(contractType: typeof(IComparable), contractName: "asdasd")] // Noncompliant
    [Export(contractType: typeof(MyInterface), contractName: "asdasd")]
    class NotExported_NamedArgs_ReverseOrder : MyInterface
    {
    }

    [Export("something", typeof(MyInterface))] // Noncompliant
    [Export(Constants.ContractName, typeof(IDisposable))] // Noncompliant
    class NotExported_MultipleArgs
    {
    }

    [Export(typeof(MyInterface)), Export(typeof(IComparable)), Export(typeof(IDisposable))]
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ {{Implement 'MyInterface' on 'NotExported_Multiple' or remove this export attribute.}}
//                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-1 {{Implement 'IComparable' on 'NotExported_Multiple' or remove this export attribute.}}
    class NotExported_Multiple : IDisposable
    {
        public void Dispose() { }
    }

    [ExportAttribute(typeof(MyInterface))] // Noncompliant
    class NotExported_FullAttributeName
    {
    }

    [Export(typeof(MyInterface))]
    [Export(typeof(Descendant))] // Noncompliant
    class Exported : MyInterface
    {
    }

    [Export]
    [Export("something")] // Exposing ourselves
    [Export(typeof(Exporting_Ourselves))]
    class Exporting_Ourselves
    {
    }

    [Export(1)]
    [Export(1, typeof(IComparable))]
    [Export(typeof(ASDASD))]
    [Export(typeof(MyInterface), typeof(IComparable))]
    class InvalidSyntax
    {
    }

    [Import(typeof(MyInterface))]
    [InheritedExport(typeof(MyInterface))] // Noncompliant
    [InheritedExport(typeof(OtherAttributes))]
    class OtherAttributes
    {
    }

    [Export(typeof(MyInterface))]
    [Export(typeof(Exported))]
    class Descendant : Exported
    {
    }

    class ExportingMembers_Are_Ignored
    {
        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyProperty { get; set; }

        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyField;

        [Export(typeof(MyInterface))]
        [Export(typeof(Exported))]
        public NotExported MyMethod() { }
    }

    interface ISomething<T> { }
    public class BaseThing { }
    public class BaseThing2 { }

    [Export(typeof(ISomething<BaseThing>))]
    public class BaseSomethingImplementation : ISomething<BaseThing>
    {

    }

    [Export(typeof(ISomething<BaseThing>))] // Noncompliant {{Implement 'ISomething<BaseThing>' on 'Something<BaseThing>' or remove this export attribute.}}
    public class Something<BaseThing>
    {
    }

    [Export(typeof(ISomething<BaseThing>))] // Noncompliant
    public class SomethingImplementation : ISomething<BaseThing2>
    {

    }

    [Export(typeof(ISomething<>))] // Noncompliant
    public class Something<T> : ISomething<T>
    {
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ShouldImplementExportedInterfaces.vb
Imports System
Imports System.ComponentModel.Composition

Namespace Classes
    Module Constants
        Public Const ContractName As String = "asdasd"
    End Module

    Interface MyInterface
    End Interface

    <Export(GetType(MyInterface))> ' Noncompliant {{Implement 'MyInterface' on 'NotExported' or remove this export attribute.}}
    <Export(GetType(Exported))> ' Noncompliant {{Derive from 'Exported' on 'NotExported' or remove this export attribute.}}
    Class NotExported
'    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-2
'    ^^^^^^^^^^^^^^^^^^^^^^^^^ @-2
    End Class

    <Export(contractType:=GetType(IComparable), contractName:="asdasd")> ' Noncompliant
    <Export(contractType:=GetType(MyInterface), contractName:="asdasd")>
    class NotExported_NamedArgs_ReverseOrder
        Implements MyInterface
    End Class

    <Export("something", GetType(MyInterface))> ' Noncompliant
    <Export(Constants.ContractName, GetType(IDisposable))> ' Noncompliant
    class NotExported_MultipleArgs
    End Class

    <Export(GetType(MyInterface)), Export(GetType(IComparable)), Export(GetType(IDisposable))>
    Class NotExported_Multiple
'    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-1 {{Implement 'MyInterface' on 'NotExported_Multiple' or remove this export attribute.}}
'                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ @-2 {{Implement 'IComparable' on 'NotExported_Multiple' or remove this export attribute.}}
        Implements IDisposable
        Public Sub Dispose()
        End Sub
    End Class

    <ExportAttribute(GetType(MyInterface))> ' Noncompliant
    Class NotExported_FullAttributeName
    End Class

    <Export(GetType(MyInterface))>
    <Export(GetType(Descendant))> ' Noncompliant
    Class Exported
        Implements MyInterface
    End Class

    <Export>
    <Export("something")> ' Exposing ourselves
    <Export(GetType(Exporting_Ourselves))>
    Class Exporting_Ourselves
    End Class

    <Export(1)>
    <Export(1, GetType(IComparable))>
    <Export(GetType(ASDASD))>
    <Export(GetType(MyInterface), GetType(IComparable))>
    Class InvalidSyntax
    End Class

    <Import(GetType(MyInterface))>
    <InheritedExport(GetType(MyInterface))> ' Noncompliant
    <InheritedExport(GetType(OtherAttributes))>
    Class OtherAttributes
    End Class

    <Export(GetType(MyInterface))>
    <Export(GetType(Exported))>
    class Descendant
        Inherits Exported
    End Class

    Class ExportingMembers_Are_Ignored
        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public Property MyProperty As NotExported

        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public MyField As NotExported

        <Export(GetType(MyInterface))>
        <Export(GetType(Exported))>
        Public Function MyMethod() As NotExported
        End Function
    End Class

    Interface ISomething(Of T)
    End Interface

    Public Class BaseThing
    End Class
    Public class BaseThing2
    End Class

    <Export(GetType(ISomething(Of BaseThing)))>
    Public Class BaseSomethingImplementation
        Implements ISomething(Of BaseThing)
    End Class

    <Export(GetType(ISomething(Of BaseThing)))> ' Noncompliant
    Public Class Something(Of BaseThing)
    End Class

    <Export(GetType(ISomething(Of BaseThing)))> ' Noncompliant
    public Class SomethingImplementation
        Implements ISomething(Of BaseThing2)
    End Class

    <Export(GetType(ISomething(Of)))> ' Noncompliant
    Public Class Something(Of T)
        Implements ISomething(Of T)
    End Class
End Namespace
